src/librbd/cache/ImageWriteback.cc

```diff
@@ -75,7 +83,7 @@ void ImageWriteback<I>::aio_flush(Context *on_finish) {
 
   auto aio_comp = io::AioCompletion::create_and_start(on_finish, &m_image_ctx,
                                                       io::AIO_TYPE_FLUSH);
-  io::ImageFlushRequest<I> req(m_image_ctx, aio_comp, io::FLUSH_SOURCE_INTERNAL,
+  io::ImageFlushRequest<I> req(m_image_ctx, aio_comp, io::FLUSH_SOURCE_WRITEBACK,
```

Comment:
Nit: probably just pass the flush type as a parameter instead of guessing.

Reply:
Guessing? This special source for the flush looped back from the ImageWriteback object at the bottom of the stack of ImageCaches avoids a deadlock.

Reply:
... but this is a generic class so if you are relying on it for your needs, it might not be valid for others.

Reply:
I understand ImageWriteBack is a write handler class which sends IO to lower layers (objecter, OSD).   During ImageWriteback<I>::aio_flush, it creates a ImageFlushRequest and sends out it . 
In a normal ImageFlushRequest, it needs to wait for the finishing of all the queued async and then completes the flush request.     
But with write back cache, a write request comes and is queued. Then it finds no cache space. At the other hand, write back cache needs to flush the former IOs, and it calls  ImageWriteBack::aio_flush. With the former logic, the aio_flush will hang and the customer IO will also hang.    
So we also update the logic in the flush request: 
https://github.com/sdpeters/ceph/blob/wip-librbd-rwl-3/src/librbd/io/ImageRequest.cc#L699-L703

Reply:
As a request, we need to differentiate whether a flush request comes from ImageWriteBack or not. This is generic for all writeback cache which use ImageWriteBack as write handler.

Reply:
Please modify the API to pass through a flush source instead of hard-coding.